### PR TITLE
Dockerization

### PR DESCRIPTION
**Pull request by rcasaleiro**
Added Dockerfile and script for importing and launching map-matcher within a container. Updated readme with docker info. Moved config.yml to map-data.

### CODE REVIEW

**Comment by bluetoothfx** [comment]
Hello, I was testing this pull request and encountered following errors.
After building successfully when I tried to run it, it says minimum memory is 8.5G. So I edited docker file and made it to 512 MB (For testing purpose). After building again it doesn't pop any error and build successfully. Now when I try to run following command as per document I has following errors. I am sharing my command. I think I am missing something, please guide me if possible :) 

`docker run --name mapmatcher -v /home/graphhoppermapmatch/map-matching/mdata:/map-matching/map-data -v /home/graphhoppermapmatch/map-matching/cdata:/map-matching/graph-cache -p 8989:8989 -e IMPORT=berlin-latest.osm.pbf -e VEHICLES="car,foot,bike" mapmatcher:master`
 

`Nothing to import, skipping...
Starting map-matcher web UI...
java.lang.IllegalArgumentException: Unable to override server.application_connectors[0].bind_host; node with index not found.
        at io.dropwizard.configuration.BaseConfigurationFactory.addOverride(BaseConfigurationFactory.java:189)
        at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:122)
        at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:89)
        at io.dropwizard.cli.ConfiguredCommand.parseConfiguration(ConfiguredCommand.java:126)
        at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:74)
        at io.dropwizard.cli.Cli.run(Cli.java:78)
        at io.dropwizard.Application.run(Application.java:93)
        at com.graphhopper.matching.http.MapMatchingApplication.main(MapMatchingApplication.java:16)
`

### If I use this with run command **IMPORT=berlin-latest.osm.pbf** then I got one more error:

`Importing file: berlin-latest.osm.pbf...
rm: cannot remove '/map-matcher/graph-cache/': Device or resource busy
java.lang.IllegalStateException: Your specified OSM file does not exist:/map-matcher/map-data/berlin-latest.osm.pbf
        at com.graphhopper.reader.osm.OSMReader.readGraph(OSMReader.java:139)
        at com.graphhopper.GraphHopper.importData(GraphHopper.java:668)
        at com.graphhopper.GraphHopper.process(GraphHopper.java:639)
        at com.graphhopper.GraphHopper.importOrLoad(GraphHopper.java:617)
        at com.graphhopper.matching.cli.ImportCommand.run(ImportCommand.java:37)
        at io.dropwizard.cli.Cli.run(Cli.java:78)
        at io.dropwizard.Application.run(Application.java:93)
        at com.graphhopper.matching.http.MapMatchingApplication.main(MapMatchingApplication.java:16)
`

**Comment by rcasaleiro** [comment]
> Hello, I was testing this pull request and encountered following errors.
> After building successfully when I tried to run it, it says minimum memory is 8.5G. So I edited docker file and made it to 512 MB (For testing purpose). After building again it doesn't pop any error and build successfully. Now when I try to run following command as per document I has following errors. I am sharing my command. I think I am missing something, please guide me if possible :)
> 
> `docker run --name mapmatcher -v /home/graphhoppermapmatch/map-matching/mdata:/map-matching/map-data -v /home/graphhoppermapmatch/map-matching/cdata:/map-matching/graph-cache -p 8989:8989 -e IMPORT=berlin-latest.osm.pbf -e VEHICLES="car,foot,bike" mapmatcher:master`
> 
> `Nothing to import, skipping... Starting map-matcher web UI... java.lang.IllegalArgumentException: Unable to override server.application_connectors[0].bind_host; node with index not found. at io.dropwizard.configuration.BaseConfigurationFactory.addOverride(BaseConfigurationFactory.java:189) at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:122) at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:89) at io.dropwizard.cli.ConfiguredCommand.parseConfiguration(ConfiguredCommand.java:126) at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:74) at io.dropwizard.cli.Cli.run(Cli.java:78) at io.dropwizard.Application.run(Application.java:93) at com.graphhopper.matching.http.MapMatchingApplication.main(MapMatchingApplication.java:16) `
> 
> ### If I use this with run command **IMPORT=berlin-latest.osm.pbf** then I got one more error:
> `Importing file: berlin-latest.osm.pbf... rm: cannot remove '/map-matcher/graph-cache/': Device or resource busy java.lang.IllegalStateException: Your specified OSM file does not exist:/map-matcher/map-data/berlin-latest.osm.pbf at com.graphhopper.reader.osm.OSMReader.readGraph(OSMReader.java:139) at com.graphhopper.GraphHopper.importData(GraphHopper.java:668) at com.graphhopper.GraphHopper.process(GraphHopper.java:639) at com.graphhopper.GraphHopper.importOrLoad(GraphHopper.java:617) at com.graphhopper.matching.cli.ImportCommand.run(ImportCommand.java:37) at io.dropwizard.cli.Cli.run(Cli.java:78) at io.dropwizard.Application.run(Application.java:93) at com.graphhopper.matching.http.MapMatchingApplication.main(MapMatchingApplication.java:16) `

**Comment by rcasaleiro** [comment]
Hello,
> After building successfully when I tried to run it, it says minimum memory is 8.5G. So I edited docker file and made it to 512 MB (For testing purpose)

If you pass "--build-arg javaopts to docker build you can change the minimum and maximum memory dynamically, there is not need to edit the Dockerfile. So you can pass --build-arg javaopts="-Xmx17g -Xms512m" to reduce minimum memory.

> Now when I try to run following command as per document I has following errors. I am sharing my command. I think I am missing something, please guide me if possible :)
> `docker run --name mapmatcher -v /home/graphhoppermapmatch/map-matching/mdata:/map-matching/map-data -v /home/graphhoppermapmatch/map-matching/cdata:/map-matching/graph-cache -p 8989:8989 -e IMPORT=berlin-latest.osm.pbf -e VEHICLES="car,foot,bike" mapmatcher:master`
> `Nothing to import, skipping... Starting map-matcher web UI... 

The command seems fine, `Nothing to import, skipping...` means that `-e IMPORT=berlin-latest.osm.pbf` above was not used. The default value `False` for IMPORT is not being overwritten. Did you add that flag to the command? If you did not, then that is expected. The last output log you sent seems to have picked up the -e IMPORT argument just fine.

> java.lang.IllegalArgumentException: Unable to override server.application_connectors[0].bind_host; node with index not found. at io.dropwizard.configuration.BaseConfigurationFactory.addOverride(BaseConfigurationFactory.java:189)

Means you already have something running on port 8989, maybe graphhopper? If you want to change this port you must do it inside config.yml and also in lines 8 and 27 of the Dockerfile.

> ### If I use this with run command **IMPORT=berlin-latest.osm.pbf** then I got one more error:
> `Importing file: berlin-latest.osm.pbf... rm: cannot remove '/map-matcher/graph-cache/': Device or resource busy java.lang.IllegalStateException: Your specified OSM file does not exist:/map-matcher/map-data/berlin-latest.osm.pbf at com.graphhopper.reader.osm.OSMReader.readGraph(OSMReader.java:139) at com.graphhopper.GraphHopper.importData(GraphHopper.java:668) at com.graphhopper.GraphHopper.process(GraphHopper.java:639) at com.graphhopper.GraphHopper.importOrLoad(GraphHopper.java:617) at com.graphhopper.matching.cli.ImportCommand.run(ImportCommand.java:37) at io.dropwizard.cli.Cli.run(Cli.java:78) at io.dropwizard.Application.run(Application.java:93) at com.graphhopper.matching.http.MapMatchingApplication.main(MapMatchingApplication.java:16) `

From this log: 
`rm: cannot remove '/map-matcher/graph-cache/': Device or resource` is expected and ignored, when the graph-cache folder does not exist.
`java.lang.IllegalStateException: Your specified OSM file does not exist:/map-matcher/map-data/berlin-latest.osm.pbf` means that you don't have that file inside `/home/graphhoppermapmatch/map-matching/mdata` which is the folder you are mounting to /map-matching/map-data (inside of the container). So make sure you copy all necessary maps and the config.yml to `/home/graphhoppermapmatch/map-matching/mdata` on your host machine.

Hope this helps!

**Comment by flashspys** [comment]
@rcasaleiro I've got a similar error

```
Importing file: koeln.pbf...
rm: cannot remove '/map-matcher/graph-cache/': Device or resource busy
Starting map-matcher web UI...
Killed
java.lang.IllegalArgumentException: Unable to override server.application_connectors[0].bind_host; node with index not found.
	at io.dropwizard.configuration.BaseConfigurationFactory.addOverride(BaseConfigurationFactory.java:189)
	at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:122)
	at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:89)
	at io.dropwizard.cli.ConfiguredCommand.parseConfiguration(ConfiguredCommand.java:126)
	at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:74)
	at io.dropwizard.cli.Cli.run(Cli.java:78)
	at io.dropwizard.Application.run(Application.java:93)
	at com.graphhopper.matching.http.MapMatchingApplication.main(MapMatchingApplication.java:16)
```
So the file exists, but the proccess got killed i guess? I've 32GB of RAM, that should not be the problem

**Comment by rcasaleiro** [comment]
Finally had some time to look at this. My bad... did not notice my configuration was different from the one in the commit... and my own lack of knowledge of Dropwizard didn't point me in the right direction. I was trying to override a configuration from the config.yml that did not exist. Everything should work fine now!

**Comment by jbordeau** [comment]
Hello,

I wanted improve this pull request, but I did not know how I could do. So I was inspired by your work to create a new pull request.
Here are the improvements:

1. Use maven-job to dockerise the application (better perf, better use docker layer etc.)
2. No need of script for the entrypoint. All arguments programs are usable
3. add docker push in the travis.yml

> Note. In fact, maybe I should have fork your repository and pull request on it. No ?

**Comment by bluetoothfx** [comment]
From my last test if I am not wrong I could not run this as docker container. @jbordeau did you mange to run it successfully ?

**Comment by jbordeau** [comment]
Yes I had the same problem. But I created a new pull request that works. You can test the image from [my docker hub](https://hub.docker.com/repository/docker/jbordeau/graphhopper-map-matching-web).

Example with a docker-compose:
`version: '3'
services:    
  graphhopper-map-matching-demo:
    image: jbordeau/graphhopper-map-matching-web
    ports:
      - "127.0.0.1:8989:8989"
    expose: 
      - "8989"
    volumes:
      - /data/graphhopper/graph-cache:/graph-cache
      - /data/graphhopper/map-data:/data
    environment:
      - JAVA_TOOL_OPTIONS="-Xmx16g"
`
The documentation and the source: https://github.com/Enovea/map-matching

In summary, you have to put a config.yml and the osm protobuf file in /data/graphhopper/map-data (if like the docker-compose above).

Or with the `docker run` command:

`docker run -v /data/graphhopper/graph-cache:/graph-cache -v /data/graphhopper/map-data:/data -e "JAVA_TOOL_OPTIONS=-Xms2g -Xmx4g" -p 8989:8989 jbordeau/graphhopper-map-matching-web`

**Comment by rcasaleiro** [comment]
> From my last test if I am not wrong I could not run this as docker container. @jbordeau did you mange to run it successfully ?

Did you try it with the new configuration file? Check my latest commit, as I had uploaded the wrong config file and didn't notice.
I haven't deployed everything from scratch in a while, but last time I tested it after that fix, it worked.

**Comment by bluetoothfx** [comment]
Hi, I have tried your solution that you told last time. It didn't work.
Meanwhile I didn't give you any feedback.

If you tell that everything is ok now, I will give another short.

Abrar

**Comment by rcasaleiro** [comment]
> Hello,
> 
> I wanted improve this pull request, but I did not know how I could do. So I was inspired by your work to create a new pull request.
> Here are the improvements:
> 
> 1. Use maven-job to dockerise the application (better perf, better use docker layer etc.)
> 2. No need of script for the entrypoint. All arguments programs are usable
> 3. add docker push in the travis.yml
> 
> > Note. In fact, maybe I should have fork your repository and pull request on it. No ?

@jbordeau thanks for the contribution! It does look neater!

**Comment by rcasaleiro** [comment]
@bluetoothfx sorry for the delay, I hope @jbordeau fixed your problem. Regardless, if you could try the latest commit I just pushed,  I just tried deploying everything from scratch and it worked fine.

**Comment by rcasaleiro** [comment]
I was deploying through docker compose and overriding the JAVA_OPTS defined in the Dockerfile, as such I wasn't having the problem before. These were the settings I just tried, with a freshly built image tagged as "mapmatcher-test:master":
`map-matcher-test:
    image: mapmatcher-test:master
    container_name: map-matcher-test
    hostname: map-matcher-test
    restart: always
    ports: 
      - 8999:8989
    volumes: 
      - ~/docker_mounts/map-matcher-test/map-data:/map-matcher/map-data
      - ~/docker_mounts/map-matcher-test/graph-cache:/map-matcher/graph-cache
    environment:
        - IMPORT=leipzig_germany.osm.pbf
        - VEHICLES=car
        - JAVA_OPTS=-server -Xconcurrentio -XX:+UseG1GC -Ddw.server.applicationConnectors[0].bindHost=0.0.0.0 -Ddw.server.applicationConnectors[0].port=8989 -Xmx20g -Xms8g
  `

**Comment by bluetoothfx** [comment]
@rcasaleiro I haven't test @jbordeau solution till now. I will test and let you guys know. Also I will check your updated solution.

**Comment by jbordeau** [comment]
@bluetoothfx you had just to use this image: https://hub.docker.com/repository/docker/jbordeau/graphhopper-map-matching-web with one command: 
`docker run -v /data/graphhopper/graph-cache:/graph-cache -v /data/graphhopper/map-data:/data -e "JAVA_TOOL_OPTIONS=-Xms2g -Xmx4g" -p 8989:8989 jbordeau/graphhopper-map-matching-web`

And that works. The import is done automatically the first time while the graph-cache exists

**Comment by bluetoothfx** [comment]
> @bluetoothfx you had just to use this image: https://hub.docker.com/repository/docker/jbordeau/graphhopper-map-matching-web with one command:
> `docker run -v /data/graphhopper/graph-cache:/graph-cache -v /data/graphhopper/map-data:/data -e "JAVA_TOOL_OPTIONS=-Xms2g -Xmx4g" -p 8989:8989 jbordeau/graphhopper-map-matching-web`
> 
> And that works. The import is done automatically the first time while the graph-cache exists

I have edited the command a bit for permission related error. The command is following:
`docker run -v ${PWD}/data/graphhopper/graph-cache:/graph-cache -v ${PWD}/data/graphhopper/map-data:/data -e "JAVA_TOOL_OPTIONS=-Xms2g -Xmx4g" -p 8989:8989 jbordeau/graphhopper-map-matching-web`

And getting error:
```
java.io.FileNotFoundException: File /data/config.yml not found
        at io.dropwizard.configuration.FileConfigurationSourceProvider.open(FileConfigurationSourceProvider.java:18)
        at io.dropwizard.configuration.BaseConfigurationFactory.build(BaseConfigurationFactory.java:79)
        at io.dropwizard.cli.ConfiguredCommand.parseConfiguration(ConfiguredCommand.java:126)
        at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:74)
        at io.dropwizard.cli.Cli.run(Cli.java:78)
        at io.dropwizard.Application.run(Application.java:94)
        at com.graphhopper.matching.http.MapMatchingApplication.main(MapMatchingApplication.java:16)
```
Am I missing something ?

**Comment by jbordeau** [comment]
indeed I forgot to specify in my last post that you must put the config.yml (graphhopper config) and the osm protobuf file in ${PWD}/data/graphhopper/map-data. These files are mandatory.

Then the command `docker run`was an example, you can change the directory volume mount up to you

**Comment by bluetoothfx** [comment]
> indeed I forgot to specify in my last post that you must put the config.yml (graphhopper config) and the osm protobuf file in ${PWD}/data/graphhopper/map-data. These files are mandatory.
> 
> Then the command `docker run`was an example, you can change the directory volume mount up to you

After putting those files in place it works like rocket. Though I have many questions. 1st how should I reference multiple proto buf files ? Multiple profile car, truck etc. And is there any docker compose for this. Please if you have time include docker-compose. In past thread I think you have made one docker-compose file click [here](https://github.com/graphhopper/map-matching/pull/173#issuecomment-683100331) but I think file need to have correct format. Please made an up and running file. Thanks. BTW in config file there is another thing called   **admin_connectors** do you have any idea about this?

Great work @jbordeau !!
